File: E_learning_frontend/E_learning_Frontend/src/app/services/Course_Service/course.service.ts
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';

@Injectable({
  providedIn: 'root'
})
export class CourseService {
  private apiUrl = 'http://localhost:8081/api/courses';

  constructor(private http :HttpClient) { }

  getAllCourses(){
    return this.http.get<any>(`${this.apiUrl}`);
  }

  getCourseByAuthor(author: string){
    return this.http.get<any>(`${this.apiUrl}/author/${author}`);
  }

  getCoursesByCategory(categoryId: string): Observable<any[]> {
    return this.http.get<any[]>(`${this.apiUrl}/category/${categoryId}`);
  }

 
  createCourse(course: any, categoryId: string): Observable<any> {
    return this.http.post<any>(`${this.apiUrl}?categoryId=${categoryId}`, course);
  }
  deleteCourse(id: string): Observable<any> {
    return this.http.delete<any>(`${this.apiUrl}/${id}`);
  }
  getCourseById(id: string): Observable<any> {
    return this.http.get<any>(`${this.apiUrl}/${id}`);
  }
  updateCourse(id: string, course: any): Observable<any> {
    return this.http.put<any>(`${this.apiUrl}/${id}`, course);
  }
}

File: E_learning_backend/src/main/java/com/Elearning/demo/MainPack/Config/CourseService.java
package com.Elearning.demo.MainPack.Config;


import com.Elearning.demo.MainPack.Model.Category;
import com.Elearning.demo.MainPack.Model.Course;
import com.Elearning.demo.MainPack.Model.CourseEnrollmentCount;
import com.Elearning.demo.MainPack.Model.Enrollment;
import com.Elearning.demo.MainPack.Repository.CourseRepository;
import com.Elearning.demo.MainPack.Repository.EnrollementRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.lang.reflect.Array;
import java.util.*;

@Service
public class CourseService {
    @Autowired
    private CategoryService categoryService;
    @Autowired
    CourseRepository courseRepository;
    @Autowired
    EnrollementRepository enrollementRepository;

    public Course createCourse(Course course, String categoryId) {
        Category category = categoryService.getCategoryById(categoryId);
        course.setCreatedDate(new Date());
        course.setUpdatedDate(new Date());
        course.setCategory(category);
        return courseRepository.save(course);
    }
    public Course updateCourse(String id, Course course) {
        Optional<Course> existingCourse = courseRepository.findById(id);
        if (existingCourse.isPresent()) {
            Course updatedCourse = existingCourse.get();
            updatedCourse.setTitle(course.getTitle());
            updatedCourse.setImageLogoUrl(course.getImageLogoUrl());
            updatedCourse.setImageUrl(course.getImageUrl());
            updatedCourse.setDescription(course.getDescription());
            updatedCourse.setAuthor(course.getAuthor());
            updatedCourse.setCategory(course.getCategory());
            updatedCourse.setUpdatedDate(new Date());
            return courseRepository.save(updatedCourse);
        } else {
            throw new RuntimeException("Course not found with id: " + id);
        }
    }

    public void deleteCourse(String id) {
        courseRepository.deleteById(id);
    }


    public Course getCourseById(String id) {
        return courseRepository.findById(id).orElseThrow(() -> new RuntimeException("Course not found with id: " + id));
    }

    public List<Course> getAllCourses() {
        return courseRepository.findAll();
    }

    public List<Course> getCoursesByAuthor(String author) {
        return courseRepository.findByAuthor(author);
    }

    public List<Course> getCoursesByCategory(Category category) {
        return courseRepository.findByCategory(category);
    }
    public List<Map<String, Object>> getCoursesWithEnrollmentCounts() {
        List<CourseEnrollmentCount> enrollmentCounts = enrollementRepository.findCourseEnrollmentCounts();
        List<Course> courses = courseRepository.findAll();

        List<Map<String, Object>> result = new ArrayList<>();
        for (Course course : courses) {
            String courseId = course.getId();
            List <Enrollment> Enrolllments  = enrollementRepository.findByCourse(course);
            Map<String, Object> courseData = new HashMap<>();
            courseData.put("courseId", course.getId());
            courseData.put("courseName", course.getTitle());
            courseData.put("enrollments", Enrolllments.size());

            result.add(courseData);
        }
        return result;
    }
}
